Implementing multi-phase grouping sets.

guofengrichard opened the pull request:
This is a work based on multi-phase aggregation in master. The idea is
to perform grouping sets aggregation in partial phase and then perform
normal aggregation in final phase.

In partial phase, we attach a GROUPINGSET_ID to each transvalue. In
final phase, we include this GROUPINGSET_ID at the head of sort keys and
group keys for group aggregation. This will ensure correctness in the
case of NULLs.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

guofengrichard commented:
This PR shows the idea of how I'd like to implement multi-phase grouping sets. There are still some todos including cost adjustment and adding test cases.

Here is a plan an an example:

```
# explain (costs off, verbose) select c1, c2, c3, avg(c3) from gstest group by grouping sets((c1,c2),(c1),(c2,c3));
                                QUERY PLAN
---------------------------------------------------------------------------
 Finalize GroupAggregate
   Output: c1, c2, c3, avg(c3)
   Group Key: (GROUPINGSET_ID()), gstest.c1, gstest.c2, gstest.c3
   ->  Sort
         Output: c1, c2, c3, (PARTIAL avg(c3)), (GROUPINGSET_ID())
         Sort Key: (GROUPINGSET_ID()), gstest.c1, gstest.c2, gstest.c3
         ->  Gather Motion 3:1  (slice1; segments: 3)
               Output: c1, c2, c3, (PARTIAL avg(c3)), (GROUPINGSET_ID())
               ->  Partial GroupAggregate
                     Output: c1, c2, c3, PARTIAL avg(c3), GROUPINGSET_ID()
                     Group Key: gstest.c1, gstest.c2
                     Group Key: gstest.c1
                     Sort Key: gstest.c2, gstest.c3
                       Group Key: gstest.c2, gstest.c3
                     ->  Sort
                           Output: c1, c2, c3
                           Sort Key: gstest.c1, gstest.c2
                           ->  Seq Scan on public.gstest
                                 Output: c1, c2, c3
 Optimizer: Postgres query optimizer
(20 rows)
```

hlinnaka commented:
Glad to see the patch is as small as this.

This needs comments. There should be a comment somewhere explaining the high-level strategy.

Eulerizeit commented:
Question, has this been tested to see if it resolves: 
https://github.com/greenplum-db/gpdb/issues/4488
https://github.com/greenplum-db/gpdb/issues/4589
https://github.com/greenplum-db/gpdb/issues/3489
https://github.com/greenplum-db/gpdb/issues/3507

Should we add those as tests for this if they have not been added?

guofengrichard commented:
> This needs comments. There should be a comment somewhere explaining the high-level strategy.

Yes, thanks. Added a comment in function `add_twostage_group_agg_path `, where the twostage grouping sets is implemented.

guofengrichard commented:
> Question, has this been tested to see if it resolves:
> #4488
Fixed.

> #4589
The wrong result is brought from upstream when merging 9.6, which has been fixed in 9.6 stable by 6520d4a9692882a7d5233dac40c4b7ff07f55049

> #3489
Fixed.

> #3507
The wrong result is brought from upstream when merging 9.6, which has been fixed in 9.6 stable by 6520d4a9692882a7d5233dac40c4b7ff07f55049

1Makers1 commented:
Can this modification solve the following issu https://github.com/greenplum-db/gpdb/issues/9156 ？
i merge the patch，but it doesn't seem to work for this issue.

create table
```
create table store_sales
(
    ss_sold_date_sk           bigint                       ,
    ss_sold_time_sk           bigint                       ,
    ss_item_sk                bigint                   ,
    ss_customer_sk            bigint                       ,
    ss_cdemo_sk               bigint                       ,
    ss_hdemo_sk               bigint                       ,
    ss_addr_sk                bigint                       ,
    ss_store_sk               bigint                       ,
    ss_promo_sk               bigint                       ,
    ss_ticket_number          bigint                   ,
    ss_quantity               bigint                       ,
    ss_wholesale_cost         decimal(7,2)                  ,
    ss_list_price             decimal(7,2)                  ,
    ss_sales_price            decimal(7,2)                  ,
    ss_ext_discount_amt       decimal(7,2)                  ,
    ss_ext_sales_price        decimal(7,2)                  ,
    ss_ext_wholesale_cost     decimal(7,2)                  ,
    ss_ext_list_price         decimal(7,2)                  ,
    ss_ext_tax                decimal(7,2)                  ,
    ss_coupon_amt             decimal(7,2)                  ,
    ss_net_paid               decimal(7,2)                  ,
    ss_net_paid_inc_tax       decimal(7,2)                  ,
    ss_net_profit             decimal(7,2)
) with (APPENDONLY=TRUE, ORIENTATION=column)  DISTRIBUTED RANDOMLY ;
create table date_dim
(
    d_date_sk                 bigint                   ,
    d_date_id                 varchar(16)                  ,
    d_date                    date                          ,
    d_month_seq               bigint                       ,
    d_week_seq                bigint                       ,
    d_quarter_seq             bigint                       ,
    d_year                    bigint                       ,
    d_dow                     bigint                       ,
    d_moy                     bigint                       ,
    d_dom                     bigint                       ,
    d_qoy                     bigint                       ,
    d_fy_year                 bigint                       ,
    d_fy_quarter_seq          bigint                       ,
    d_fy_week_seq             bigint                       ,
    d_day_name                varchar(9)                       ,
    d_quarter_name            varchar(6)                       ,
    d_holiday                 varchar(1)                       ,
    d_weekend                 varchar(1)                       ,
    d_following_holiday       varchar(1)                       ,
    d_first_dom               bigint                       ,
    d_last_dom                bigint                       ,
    d_same_day_ly             bigint                       ,
    d_same_day_lq             bigint                       ,
    d_current_day             varchar(1)                       ,
    d_current_week            varchar(1)                       ,
    d_current_month           varchar(1)                       ,
    d_current_quarter         varchar(1)                       ,
    d_current_year            varchar(1)
) with (APPENDONLY=TRUE, ORIENTATION=column)  DISTRIBUTED RANDOMLY ;
create table store
(
    s_store_sk                bigint                   ,
    s_store_id                varchar(16)                  ,
    s_rec_start_date          date                          ,
    s_rec_end_date            date                          ,
    s_closed_date_sk          bigint                       ,
    s_store_name              varchar(50)                   ,
    s_number_employees        bigint                       ,
    s_floor_space             bigint                       ,
    s_hours                   varchar(20)                      ,
    s_manager                 varchar(40)                   ,
    s_market_id               bigint                       ,
    s_geography_class         varchar(100)                  ,
    s_market_desc             varchar(100)                  ,
    s_market_manager          varchar(40)                   ,
    s_division_id             bigint                       ,
    s_division_name           varchar(50)                   ,
    s_company_id              bigint                       ,
    s_company_name            varchar(50)                   ,
    s_street_number           varchar(10)                   ,
    s_street_name             varchar(60)                   ,
    s_street_type             varchar(15)                      ,
    s_suite_number            varchar(10)                      ,
    s_city                    varchar(60)                   ,
    s_county                  varchar(30)                   ,
    s_state                   varchar(2)                       ,
    s_zip                     varchar(10)                      ,
    s_country                 varchar(20)                   ,
    s_gmt_offset              decimal(5,2)                  ,
    s_tax_precentage          decimal(5,2)
) with (APPENDONLY=TRUE, ORIENTATION=column)  DISTRIBUTED RANDOMLY ;
create table item
(
    i_item_sk                 bigint                   ,
    i_item_id                 varchar(16)                  ,
    i_rec_start_date          date                          ,
    i_rec_end_date            date                          ,
    i_item_desc               varchar(200)                  ,
    i_current_price           decimal(7,2)                  ,
    i_wholesale_cost          decimal(7,2)                  ,
    i_brand_id                bigint                       ,
    i_brand                   varchar(50)                      ,
    i_class_id                bigint                       ,
    i_class                   varchar(50)                      ,
    i_category_id             bigint                       ,
    i_category                varchar(50)                      ,
    i_manufact_id             bigint                       ,
    i_manufact                varchar(50)                      ,
    i_size                    varchar(20)                      ,
    i_formulation             varchar(20)                      ,
    i_color                   varchar(20)                      ,
    i_units                   varchar(10)                      ,
    i_container               varchar(10)                      ,
    i_manager_id              bigint                       ,
    i_product_name            varchar(50)
) with (APPENDONLY=TRUE, ORIENTATION=column)  DISTRIBUTED RANDOMLY ;
```
loader tpcds 1g data

Observe the execution plan

```
explain analyze select  *
from (select i_category
            ,i_class
            ,i_brand
            ,i_product_name
            ,d_year
            ,d_qoy
            ,d_moy
            ,s_store_id
            ,sumsales
            ,rank() over (partition by i_category order by sumsales desc) rk
      from (select i_category
                  ,i_class
                  ,i_brand
                  ,i_product_name
                  ,d_year
                  ,d_qoy
                  ,d_moy
                  ,s_store_id
                  ,sum(coalesce(ss_sales_price*ss_quantity,0)) sumsales
            from store_sales
                ,date_dim
                ,store
                ,item
       where  ss_sold_date_sk=d_date_sk
          and ss_item_sk=i_item_sk
          and ss_store_sk = s_store_sk
          and d_month_seq between 1201 and 1201+11
       group by  rollup(i_category, i_class, i_brand, i_product_name, d_year, d_qoy, d_moy,s_store_id))dw1) dw2
where rk <= 100
order by i_category
        ,i_class
        ,i_brand
        ,i_product_name
        ,d_year
        ,d_qoy
        ,d_moy
        ,s_store_id
        ,sumsales
        ,rk
limit 100;
                                                                                                      QUERY PLAN

------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
----------------
 Limit  (cost=64364.14..64364.39 rows=100 width=131) (actual time=13800.586..13800.655 rows=100 loops=1)
   ->  Sort  (cost=64364.14..64415.46 rows=20530 width=131) (actual time=13800.534..13800.544 rows=100 loops=1)
         Sort Key: dw2.i_category, dw2.i_class, dw2.i_brand, dw2.i_product_name, dw2.d_year, dw2.d_qoy, dw2.d_moy, dw2.s_store_id, dw2.sumsales, dw2.rk
         Sort Method:  top-N heapsort  Memory: 57kB
         ->  Subquery Scan on dw2  (cost=61577.94..63579.53 rows=20530 width=131) (actual time=12746.262..13798.092 rows=1100 loops=1)
               Filter: (dw2.rk <= 100)
               Rows Removed by Filter: 519742
               ->  WindowAgg  (cost=61577.94..62809.69 rows=61588 width=131) (actual time=12746.220..13701.795 rows=520842 loops=1)
                     Partition By: dw1.i_category
                     Order By: dw1.sumsales
                     ->  Sort  (cost=61577.94..61731.91 rows=61588 width=123) (actual time=12746.160..12919.120 rows=520842 loops=1)
                           Sort Key: dw1.i_category, dw1.sumsales DESC
                           Sort Method:  external sort  Disk: 58656kB
                           ->  Subquery Scan on dw1  (cost=54704.98..56678.56 rows=61588 width=123) (actual time=4781.165..8312.189 rows=520842 loops=1)
                                 ->  GroupAggregate  (cost=54704.98..56062.69 rows=61588 width=123) (actual time=4781.157..8166.925 rows=520842 loops=1)
                                       Group Key: item.i_category, item.i_class, item.i_brand, item.i_product_name, date_dim.d_year, date_dim.d_qoy, date_dim.d_moy, store.s_store_id
                                       Group Key: item.i_category, item.i_class, item.i_brand, item.i_product_name, date_dim.d_year, date_dim.d_qoy, date_dim.d_moy
                                       Group Key: item.i_category, item.i_class, item.i_brand, item.i_product_name, date_dim.d_year, date_dim.d_qoy
                                       Group Key: item.i_category, item.i_class, item.i_brand, item.i_product_name, date_dim.d_year
                                       Group Key: item.i_category, item.i_class, item.i_brand, item.i_product_name
                                       Group Key: item.i_category, item.i_class, item.i_brand
                                       Group Key: item.i_category, item.i_class
                                       Group Key: item.i_category
                                       Group Key: ()
                                       ->  Gather Motion 3:1  (slice5; segments: 3)  (cost=54704.98..54969.52 rows=11758 width=105) (actual time=4781.093..6437.962 rows=538691 loops=1)
                                             Merge Key: item.i_category, item.i_class, item.i_brand, item.i_product_name, date_dim.d_year, date_dim.d_qoy, date_dim.d_moy, store.s_store_id
                                             ->  Sort  (cost=54704.98..54734.37 rows=3920 width=105) (actual time=4702.081..5908.106 rows=182472 loops=1)
                                                   Sort Key: item.i_category, item.i_class, item.i_brand, item.i_product_name, date_dim.d_year, date_dim.d_qoy, date_dim.d_moy, store.s_store_id
                                                   Sort Method:  external merge  Disk: 69024kB
                                                   ->  Hash Join  (cost=2456.22..53910.11 rows=3920 width=105) (actual time=87.659..2073.317 rows=182472 loops=1)
                                                         Hash Cond: (store_sales.ss_item_sk = item.i_item_sk)
                                                         Extra Text: (seg1)   Hash chain length 1.0 avg, 3 max, using 5832 of 65536 buckets.
                                                         ->  Redistribute Motion 3:3  (slice3; segments: 3)  (cost=1526.22..52818.45 rows=3920 width=63) (actual time=0.062..1777.846 rows=182472 loop
s=1)
                                                               Hash Key: store_sales.ss_item_sk
                                                               ->  Hash Join  (cost=1526.22..52583.31 rows=3920 width=63) (actual time=25.059..1655.571 rows=180104 loops=1)
                                                                     Hash Cond: (store_sales.ss_store_sk = store.s_store_sk)
                                                                     Extra Text: (seg1)   Hash chain length 1.0 avg, 1 max, using 12 of 65536 buckets.
                                                                     ->  Hash Join  (cost=1524.17..52417.44 rows=4111 width=54) (actual time=16.311..1533.992 rows=184505 loops=1)
                                                                           Hash Cond: (store_sales.ss_sold_date_sk = date_dim.d_date_sk)
                                                                           Extra Text: (seg1)   Hash chain length 1.0 avg, 2 max, using 365 of 65536 buckets.
                                                                           ->  Seq Scan on store_sales  (cost=0.00..42917.04 rows=960135 width=38) (actual time=2.978..1204.137 rows=961942 loops=1)
                                                                           ->  Hash  (cost=1511.86..1511.86 rows=329 width=32) (actual time=11.439..11.439 rows=366 loops=1)
                                                                                 ->  Broadcast Motion 3:3  (slice1; segments: 3)  (cost=0.00..1511.86 rows=329 width=32) (actual time=0.046..11.279 ro
ws=366 loops=1)
                                                                                       ->  Seq Scan on date_dim  (cost=0.00..1498.73 rows=110 width=32) (actual time=4.758..10.232 rows=131 loops=1)
                                                                                             Filter: ((d_month_seq >= 1201) AND (d_month_seq <= 1212))
                                                                     ->  Hash  (cost=1.60..1.60 rows=12 width=25) (actual time=4.411..4.411 rows=12 loops=1)
                                                                           ->  Broadcast Motion 3:3  (slice2; segments: 3)  (cost=0.00..1.60 rows=12 width=25) (actual time=4.369..4.377 rows=12 loops
=1)
                                                                                 ->  Seq Scan on store  (cost=0.00..1.12 rows=4 width=25) (actual time=1.681..1.697 rows=5 loops=1)
                                                         ->  Hash  (cost=705.00..705.00 rows=6000 width=58) (actual time=75.633..75.633 rows=6106 loops=1)
                                                               ->  Redistribute Motion 3:3  (slice4; segments: 3)  (cost=0.00..705.00 rows=6000 width=58) (actual time=0.034..73.452 rows=6106 loops=1
)
                                                                     Hash Key: item.i_item_sk
                                                                     ->  Seq Scan on item  (cost=0.00..345.00 rows=6000 width=58) (actual time=0.931..3.583 rows=6105 loops=1)
 Planning time: 3.416 ms
 * (slice0)    Executor memory: 64472K bytes.  Work_mem: 63624K bytes max, 131784K bytes wanted.
   (slice1)    Executor memory: 592K bytes avg x 3 workers, 592K bytes max (seg0).
   (slice2)    Executor memory: 300K bytes avg x 3 workers, 300K bytes max (seg0).
   (slice3)    Executor memory: 3689K bytes avg x 3 workers, 3689K bytes max (seg0).  Work_mem: 533K bytes max.
   (slice4)    Executor memory: 592K bytes avg x 3 workers, 592K bytes max (seg0).
 * (slice5)    Executor memory: 46491K bytes avg x 3 workers, 46494K bytes max (seg0).  Work_mem: 45269K bytes max, 61750K bytes wanted.
 Memory used:  128000kB
 Memory wanted:  924084kB
 Optimizer: Postgres query optimizer
 Execution time: 13880.551 ms
```

guofengrichard commented:
> Can this modification solve the following issu #9156 ？
> i merge the patch，but it doesn't seem to work for this issue.

In your case the two-stage path is generated and then fails when competing with the one-stage path in add_path(). I think this is reasonable because we always generate both paths and pick the cheaper one.

1Makers1 commented:
For the code which i merge the patch, I used a larger amount of data to test. it seems to 
spend a lot of time on the master. If they run on the segment instances, they can perform operations in parallel, which costs may be more expensive, but because it is multiple parallel operations, it may take less time.

```
postgres=# \d+
                                     List of relations
 Schema |      Name       | Type  |  Owner  |       Storage        |  Size   | Description
--------+-----------------+-------+---------+----------------------+---------+-------------
 public | date_dim        | table | ssadmin | append only columnar | 16 MB   |
 public | ext_date_dim    | table | ssadmin | external             | 0 bytes |
 public | ext_item        | table | ssadmin | external             | 0 bytes |
 public | ext_store       | table | ssadmin | external             | 0 bytes |
 public | ext_store_sales | table | ssadmin | external             | 0 bytes |
 public | item            | table | ssadmin | append only columnar | 90 MB   |
 public | store           | table | ssadmin | append only columnar | 3560 kB |
 public | store_sales     | table | ssadmin | append only columnar | 172 GB  |
 public | time_dim        | table | ssadmin | append only columnar | 544 kB  |

```
```
postgres=# explain analyze select  *
from (select i_category
            ,i_class
            ,i_brand
            ,i_product_name
            ,d_year
            ,d_qoy
            ,d_moy
            ,s_store_id
            ,sumsales
            ,rank() over (partition by i_category order by sumsales desc) rk
      from (select i_category
                  ,i_class
                  ,i_brand
                  ,i_product_name
                  ,d_year
                  ,d_qoy
                  ,d_moy
                  ,s_store_id
                  ,sum(coalesce(ss_sales_price*ss_quantity,0)) sumsales
            from store_sales
                ,date_dim
                ,store
                ,item
       where  ss_sold_date_sk=d_date_sk
          and ss_item_sk=i_item_sk
          and ss_store_sk = s_store_sk
          and d_month_seq between 1220 and 1220+11
       group by  rollup(i_category, i_class, i_brand, i_product_name, d_year, d_qoy, d_moy,s_store_id))dw1) dw2
where rk <= 100
order by i_category
        ,i_class
        ,i_brand
        ,i_product_name
        ,d_year
        ,d_qoy
        ,d_moy
        ,s_store_id
        ,sumsales
        ,rk
limit 100;
                                                                                                         QUERY PLAN

---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
-------------
 Limit  (cost=31362792.43..31362792.68 rows=100 width=135) (actual time=2752427.406..2752427.436 rows=100 loops=1)
   ->  Sort  (cost=31362792.43..31378184.97 rows=6157014 width=135) (actual time=2752427.403..2752427.408 rows=100 loops=1)
         Sort Key: dw2.i_category, dw2.i_class, dw2.i_brand, dw2.i_product_name, dw2.d_year, dw2.d_qoy, dw2.d_moy, dw2.s_store_id, dw2.sumsales, dw2.rk
         Sort Method:  top-N heapsort  Memory: 57kB
         ->  Subquery Scan on dw2  (cost=30527166.94..31127475.79 rows=6157014 width=135) (actual time=2516734.701..2752426.150 rows=1100 loops=1)
               Filter: (dw2.rk <= 100)
               Rows Removed by Filter: 189384453
               ->  WindowAgg  (cost=30527166.94..30896587.77 rows=18471042 width=135) (actual time=2516734.688..2740292.476 rows=189385553 loops=1)
                     Partition By: dw1.i_category
                     Order By: dw1.sumsales
                     ->  Sort  (cost=30527166.94..30573344.54 rows=18471042 width=127) (actual time=2516734.647..2625528.304 rows=189385553 loops=1)
                           Sort Key: dw1.i_category, dw1.sumsales DESC
                           Sort Method:  external merge  Disk: 23640352kB
                           ->  Subquery Scan on dw1  (cost=23242844.86..23885203.55 rows=18471042 width=127) (actual time=211042.846..1362189.437 rows=189385553 loops=1)
                                 ->  GroupAggregate  (cost=23242844.86..23700493.14 rows=18471042 width=127) (actual time=211042.844..1341156.493 rows=189385553 loops=1)
                                       Group Key: item.i_category, item.i_class, item.i_brand, item.i_product_name, date_dim.d_year, date_dim.d_qoy, date_dim.d_moy, store.s_store_id
                                       Group Key: item.i_category, item.i_class, item.i_brand, item.i_product_name, date_dim.d_year, date_dim.d_qoy, date_dim.d_moy
                                       Group Key: item.i_category, item.i_class, item.i_brand, item.i_product_name, date_dim.d_year, date_dim.d_qoy
                                       Group Key: item.i_category, item.i_class, item.i_brand, item.i_product_name, date_dim.d_year
                                       Group Key: item.i_category, item.i_class, item.i_brand, item.i_product_name
                                       Group Key: item.i_category, item.i_class, item.i_brand
                                       Group Key: item.i_category, item.i_class
                                       Group Key: item.i_category
                                       Group Key: ()
                                       ->  Gather Motion 16:1  (slice5; segments: 16)  (cost=23242844.86..23344886.97 rows=4535206 width=109) (actual time=211042.763..1001061.274 rows=214873679 loops=1)
                                             Merge Key: item.i_category, item.i_class, item.i_brand, item.i_product_name, date_dim.d_year, date_dim.d_qoy, date_dim.d_moy, store.s_store_id
                                             ->  Sort  (cost=23242844.86..23254182.87 rows=283451 width=109) (actual time=210885.646..302748.167 rows=13627193 loops=1)
                                                   Sort Key: item.i_category, item.i_class, item.i_brand, item.i_product_name, date_dim.d_year, date_dim.d_qoy, date_dim.d_moy, store.s_store_id
                                                   Sort Method:  external merge  Disk: 28562336kB
                                                   ->  Hash Join  (cost=17545.73..21772031.38 rows=283451 width=109) (actual time=39.474..35087.702 rows=13627193 loops=1)
                                                         Hash Cond: (store_sales.ss_store_sk = store.s_store_sk)
                                                         Extra Text: (seg7)   Hash chain length 1.0 avg, 2 max, using 997 of 65536 buckets.
                                                         ->  Hash Join  (cost=17154.97..21708548.28 rows=295672 width=100) (actual time=38.259..29481.652 rows=13955708 loops=1)
                                                               Hash Cond: (store_sales.ss_item_sk = item.i_item_sk)
                                                               Extra Text: (seg7)   Hash chain length 1.1 avg, 4 max, using 16376 of 65536 buckets.
                                                               ->  Redistribute Motion 16:16  (slice2; segments: 16)  (cost=1614.97..21627960.57 rows=295672 width=54) (actual time=0.023..18317.746 rows=13955
708 loops=1)
                                                                     Hash Key: store_sales.ss_item_sk
                                                                     ->  Hash Join  (cost=1614.97..21533345.72 rows=295672 width=54) (actual time=6.666..75249.713 rows=13758386 loops=1)
                                                                           Hash Cond: (store_sales.ss_sold_date_sk = date_dim.d_date_sk)
                                                                           Extra Text: (seg3)   Hash chain length 1.0 avg, 2 max, using 364 of 65536 buckets.
                                                                           ->  Seq Scan on store_sales  (cost=0.00..17164467.96 rows=71999256 width=38) (actual time=2.227..55759.101 rows=72012982 loops=1)
                                                                           ->  Hash  (cost=1552.14..1552.14 rows=315 width=32) (actual time=2.937..2.937 rows=365 loops=1)
                                                                                 ->  Broadcast Motion 16:16  (slice1; segments: 16)  (cost=0.00..1552.14 rows=315 width=32) (actual time=0.020..2.809 rows=365
loops=1)
                                                                                       ->  Seq Scan on date_dim  (cost=0.00..1498.73 rows=20 width=32) (actual time=1.580..2.281 rows=32 loops=1)
                                                                                             Filter: ((d_month_seq >= 1220) AND (d_month_seq <= 1231))
                                                               ->  Hash  (cost=11790.00..11790.00 rows=18750 width=62) (actual time=37.950..37.950 rows=18969 loops=1)
                                                                     ->  Redistribute Motion 16:16  (slice3; segments: 16)  (cost=0.00..11790.00 rows=18750 width=62) (actual time=0.038..31.319 rows=18969 loo
ps=1)
                                                                           Hash Key: item.i_item_sk
                                                                           ->  Seq Scan on item  (cost=0.00..5790.00 rows=18750 width=62) (actual time=1.006..8.672 rows=18984 loops=1)
                                                         ->  Hash  (cost=190.36..190.36 rows=1002 width=25) (actual time=2.340..2.340 rows=1002 loops=1)
                                                               ->  Broadcast Motion 16:16  (slice4; segments: 16)  (cost=0.00..190.36 rows=1002 width=25) (actual time=0.057..1.838 rows=1002 loops=1)
                                                                     ->  Seq Scan on store  (cost=0.00..20.02 rows=63 width=25) (actual time=1.036..1.083 rows=76 loops=1)
 Planning time: 3.040 ms
 * (slice0)    Executor memory: 131077K bytes.  Work_mem: 129925K bytes max, 41680343K bytes wanted.
   (slice1)    Executor memory: 592K bytes avg x 16 workers, 592K bytes max (seg0).
   (slice2)    Executor memory: 2969K bytes avg x 16 workers, 2969K bytes max (seg0).  Work_mem: 532K bytes max.
   (slice3)    Executor memory: 592K bytes avg x 16 workers, 592K bytes max (seg0).
   (slice4)    Executor memory: 300K bytes avg x 16 workers, 300K bytes max (seg0).
 * (slice5)    Executor memory: 47886K bytes avg x 16 workers, 47920K bytes max (seg11).  Work_mem: 44724K bytes max, 5430915K bytes wanted.
 Memory used:  128000kB
 Memory wanted:  291763994kB
 Optimizer: Postgres query optimizer
 Execution time: 2752491.350 ms
(63 rows)

```

1Makers1 commented:
Can you give me more suggestions and ideas, so that I can do more thing and give more information

guofengrichard commented:
> Can you give me more suggestions and ideas, so that I can do more thing and give more information

Would you please provide the way you loaded data so I can reproduce your issue myself?

1Makers1 commented:
1.download the tool which can produce data
2.compile the tool
3.produce data
4.load the data

for example

**1.download the tool which can produce data**
Enter the tpcds official website [tpcds](http://www.tpc.org/tpc_documents_current_versions/current_specifications5.asp) ，and find a way to download the [ tpcds tool](http://www.tpc.org/TPC_Documents_Current_Versions/download_programs/tools-download-request5.asp?bm_type=TPC-DS&bm_vers=2.11.0&mode=CURRENT-ONLY)

**2.compile the tool**
unzip the package，enter the directory and compile
`cd /root/v2.10.1rc3/tools ; make`

**3.produce data**
then produce data (this mean produce data in /opt/tpcds/1g  and SCALE is 1GB)
`./dsdgen -DIR /opt/tpcds/1g  -SCALE 1  -TERMINATE  N `
if you want to produce a large mount of data. you can use the following command to generate data
in paralle.
(this mean produce data in /opt/tpcds/100g  and SCALE is 100GB USING 4 processes  .  -PARALLEL mean how many processes total.  -CHILD means which part number of data it will produce.)
```
./dsdgen -DIR /opt/tpcds/100g  -SCALE 100 -PARALLEL 4 -CHILD 1 -TERMINATE N
./dsdgen -DIR /opt/tpcds/100g  -SCALE 100 -PARALLEL 4 -CHILD 2 -TERMINATE N
./dsdgen -DIR /opt/tpcds/100g  -SCALE 100 -PARALLEL 4 -CHILD 3 -TERMINATE N
./dsdgen -DIR /opt/tpcds/100g  -SCALE 100 -PARALLEL 4 -CHILD 4 -TERMINATE N

```

**4.load the data**
it will produce data like this，and then you can loader the data use copy or gpfdist
```

[root@node1 tools]# ll /opt/tpcds/1g/
total 1204832
-rw-r--r-- 1 root root  13109372 Jan  2 02:21 customer.dat
-rw-r--r-- 1 root root   5452165 Jan  2 02:21 customer_address.dat
-rw-r--r-- 1 root root  78739296 Jan  2 02:21 customer_demographics.dat
-rw-r--r-- 1 root root  10244389 Jan  2 02:21 date_dim.dat
-rw-r--r-- 1 root root   5033899 Jan  2 02:21 item.dat
-rw-r--r-- 1 root root      3143 Jan  2 02:21 store.dat
-rw-r--r-- 1 root root  32422491 Jan  2 02:22 store_returns.dat
-rw-r--r-- 1 root root 385565005 Jan  2 02:22 store_sales.dat
```

guofengrichard commented:
> For the code which i merge the patch, I used a larger amount of data to test. it seems to
> spend a lot of time on the master. If they run on the segment instances, they can perform operations in parallel, which costs may be more expensive, but because it is multiple parallel operations, it may take less time.

Cheaper cost is supposed to indicate less execution time, if the cost estimation does not go wrong.

I have checked your case, with the same DDL and the same data. In my env, the execution time of one-phase grouping sets is `Execution time: 16230.353 ms`, while for two-phase grouping sets it is `Execution time: 23081.056 ms`. And as we can see, the planner chooses the one-phase grouping sets as its final plan. I think it's doing the right thing.

1Makers1 commented:
i use the data size of 1TB. Can you describe the amount of your data and show your execution plan？and if there any way for me to  test one-phase grouping sets and two-phase grouping sets .thanks

guofengrichard commented:
> i use the data size of 1TB. Can you describe the amount of your data and show your execution plan？and if there any way for me to test one-phase grouping sets and two-phase grouping sets .thanks

You can do a little hack on this patch:

```
--- a/src/backend/optimizer/plan/planner.c
+++ b/src/backend/optimizer/plan/planner.c
@@ -4557,6 +4557,8 @@ create_grouping_paths(PlannerInfo *root,
      * Add GPDB two-and three-stage agg plans
      */
     if (try_mpp_multistage_aggregation)
+    {
+        grouped_rel->pathlist = NIL;
         cdb_create_twostage_grouping_paths(root,
                                            input_rel,
                                            grouped_rel,
@@ -4570,6 +4572,7 @@ create_grouping_paths(PlannerInfo *root,
                                            &agg_final_costs,
                                            rollup_lists,
                                            rollup_groupclauses);
+    }
```

Then you can generate one-stage/two-stage plan by setting GUC `gp_enable_multiphase_agg` to off/on.

hlinnaka commented:
With this patch, does the planner now have the capability to produce all the same plan types as before the 9.5 merge? Or different plans, but with similar performance. Hashed aggregrates are still not supported, at least.

guofengrichard commented:
> I was expecting more test cases with this. We have a lot of GROUPING SETS queries in the regression suite, how many of the existing tests pick a two-stage plan now? I'd suggest adding some cases with EXPLAIN, to specifically check when two-stage plans are chosen or not chosen. Maybe in the `olap_plans` test, I added that kind of tests for GROUP BY without grouping sets there, as part of the 9.6 merge.

Added more test cases in `olap_plans` to demonstrate when two-stage plans are chosen or not chosen. Basically when the query produces a relatively small number of groups in comparison to the number of input rows, two-stage plans tend to be picked.

I'm not sure how many existing grouping sets queries would use two-stage plan. I checked test case `groupingset` and most queries do not have EXPLAIN.

guofengrichard commented on src/backend/cdb/cdbgroupingpaths.c:
```diff
@@ -187,13 +254,87 @@ add_twostage_group_agg_path(PlannerInfo *root,
+			add_gsetid_groupclause(copyObject(parse->groupClause), groupref);
+
+		if (!list_member(ctx->partial_grouping_target->exprs, gsetid))
+			add_column_to_pathtarget(ctx->partial_grouping_target,
```
Currently we will not use this PathTarget to create other kind of plan. Not sure if this holds true for the future. Agree that better safe than sorry. Added the copy codes.

guofengrichard commented on src/backend/executor/nodeAgg.c:
```diff
@@ -2507,16 +2512,11 @@ ExecInitAgg(Agg *node, EState *estate, int eflags)
 		phasedata->sortnode = sortnode;
 
 		/* Compute group_ids */
+		phasedata->group_id = palloc0(numGroupingSets * sizeof(int));
```
Yes, that makes sense. Will open a separate PR for this refactoring.

guofengrichard commented on src/include/nodes/primnodes.h:
```diff
@@ -366,6 +366,16 @@ typedef struct GroupId
+typedef struct GroupingSetId
+{
+	Expr		xpr;
+	Index		agglevelsup;	/* same as Aggref.agglevelsup */
```
Removed. Thanks.

hlinnaka commented:
On 17 January 2020 10:42:14 EET, Richard Guo <notifications@github.com> wrote:
>I'm not sure how many existing grouping sets queries would use
>two-stage plan. I checked test case `groupingset` and most queries do
>not have EXPLAIN.

Maybe add an elog(NOTICE) to the place where the new GROUPINGSET_ID() expression is evaluated, and run the regression tests locally.


- Heikki

guofengrichard commented:
> With this patch, does the planner now have the capability to produce all the same plan types as before the 9.5 merge? Or different plans, but with similar performance. Hashed aggregrates are still not supported, at least.

Before 9.5 merge GPDB performs grouping sets in a different way. It can use HashAggregate (as you mentioned) and also it can redistribute the partial results to all segments so the final aggregation can be performed in parallel.  So it would have better performance than now. An example as below:

```
# explain (costs off) select a, b, sum(d) from olap_test group by grouping sets((a, b), (a));
                                              QUERY PLAN
-------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice2; segments: 3)
   ->  HashAggregate
         Group Key: "rollup".a, "rollup".b, "rollup"."grouping", "rollup"."group_id"
         ->  Subquery Scan on "rollup"
               ->  Redistribute Motion 3:3  (slice1; segments: 3)
                     Hash Key: rollup_1.a, rollup_1.b, (Grouping), group_id()
                     ->  GroupAggregate
                           Group Key: rollup_1.a, rollup_1."grouping", rollup_1."group_id", rollup_1.b
                           ->  Subquery Scan on rollup_1
                                 ->  GroupAggregate
                                       Group Key: olap_test.a, olap_test.b
                                       ->  Sort
                                             Sort Key: olap_test.a, olap_test.b
                                             ->  Seq Scan on olap_test
 Optimizer: Postgres query optimizer
(15 rows)
```

For the HashAggregate, we can get it when merging PG 10.
We may need to add a path with Redistribute Motion as next step.

guofengrichard commented:
> On 17 January 2020 10:42:14 EET, Richard Guo ***@***.***> wrote: I'm not sure how many existing grouping sets queries would use two-stage plan. I checked test case `groupingset` and most queries do not have EXPLAIN.
> Maybe add an elog(NOTICE) to the place where the new GROUPINGSET_ID() expression is evaluated, and run the regression tests locally. - Heikki

Aha, that's smart. Will do.

guofengrichard commented:
Thank you all for reviewing this PR. I have merged it into master.

(The `Reviewed-by` was not added successfully in the commit message. Feeling sorry for that :( )
